lib/opte/src/ddi/sync.rs

```diff
@@ -409,3 +427,327 @@ impl<T> KRwLock<T> {
+                };
+                (lock, cause)
+            } else {
+                let (lock, cause) = self.wait_timeout(lock, dur);
+
+                let cause = match cause {
+                    TimedWaitCause::Notify => TimedWaitSigCause::Notify,
+                    TimedWaitCause::Timeout => TimedWaitSigCause::Timeout,
+                };
+
+                (lock, cause)
+            }
+        }
+    }
+}
+
+#[cfg(all(not(feature = "std"), not(test)))]
+impl Drop for KCondvar {
+    fn drop(&mut self) {
+        unsafe { cv_destroy(self.cv.get()) };
+    }
+}
+
+pub enum WaitSigCause {
+    Notify,
+    Signal,
+}
+
+pub enum TimedWaitCause {
+    Notify,
+    Timeout,
+}
+
+pub enum TimedWaitSigCause {
+    Notify,
+    Signal,
+    Timeout,
+}
+
+/// A mutual exclusion mechanism which loans out access to a single
+/// internal token. This is used to ensure at most one thread is present
+/// in a critical section *without actively holding a [`KMutex`]*.
+///
+/// This is necessary for some kernel-level operations which will upcall or
+/// enter some other context in which it is unsafe to hold a [`KMutex`] or
+/// [`KRwLock`]. Any functions attached to the token `T` should denote whether
+/// these restrictions apply.
+pub struct TokenLock<T> {
+    // In future, this could be arbitrary (i.e., taking user-held
+    // context like a file descriptor). Similarly this could allow
+    // for re-entrancy, but this is tricky to square with `&mut` access
+    // to the inner `T`.
+    #[cfg(all(not(feature = "std"), not(test)))]
+    holder: KMutex<Option<NonNull<kthread_t>>>,
+    #[cfg(any(feature = "std", test))]
+    holder: KMutex<Option<ThreadId>>,
+    cv: KCondvar,
+    inner: UnsafeCell<T>,
+}
+
+impl<T> TokenLock<T> {
+    pub fn new(token: T) -> Self {
+        let holder = KMutex::new(None);
+        let cv = KCondvar::new();
+
+        Self { holder, cv, inner: UnsafeCell::new(token) }
+    }
+
+    pub fn lock(&self) -> Token<'_, T> {
+        let mut thread_lock = self.holder.lock();
+
+        while thread_lock.is_some() {
+            // Here, we expect that there is a very real possibility
+            // that the kthread in `thread_lock` is, itself, susceptible
+            // to being `STOPPED` (and thus deadlocking us) if we cannot
+            // also handle a SIGSTOP. We don't care whether we were awoken
+            // by the signal or the CV, only that our thread can also stop
+            // to allow a fork of the caller to complete.
+            thread_lock = self.cv.wait_sig(thread_lock).0;
+        }
+
+        #[cfg(all(not(feature = "std"), not(test)))]
+        let curthread = unsafe {
+            NonNull::new(threadp())
+                .expect("current thread *must* be a valid pointer")
+        };
+
+        #[cfg(any(feature = "std", test))]
+        let curthread = std::thread::current().id();
+
+        *thread_lock = Some(curthread);
```

Comment:
This differs slightly from other times this pattern occurs in the kernel (e.g., nvme.c) -- here we drop the lock as soon as possible and stake our claim purely using the thread ID, whereas in the prior art the thread ID is only stored when making an upcall.

If having the same behaviour is crucial, we could shift the thread ID logic into a `Token::upcall_context(&mut self, f: impl FnOnce())` and have `Token` hold a lock guard instead -- this would also be useful in proving that none of the inner data/locks are held in a given closure.